src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockImporter.scala

```diff
@@ -0,0 +1,270 @@
+          log.debug(s"Ignoring orphaned block $newNumber ($headerHash) from $peerId")
+
+        case ChainReorganised(oldBranch, newBranch, totalDifficulties) =>
+          updateTxAndOmmerPools(newBranch, oldBranch)
+          broadcastBlocks(newBranch, totalDifficulties)
+          val header = newBranch.last.header
+          log.debug(
+            s"Imported block $newNumber ($headerHash) from $peerId, " +
+              s"resulting in chain reorganisation: new branch of length ${newBranch.size} with head at block " +
+              s"${header.number} (${hash2string(header.hash)})")
+
+        case BlockImportFailed(error) =>
+          fetcher ! BlockFetcher.BlockImportFailed(Block.number(block), error)
+      }
+      .map(_ => running _)
+      .recover {
+        case missingNodeEx: MissingNodeException if syncConfig.redownloadMissingStateNodes =>
+          // state node re-download will be handled when downloading headers
+          log.error("Ignoring broadcast block, reason: {}", missingNodeEx)
+          running
+      }
+  })
+
+  private def broadcastBlocks(blocks: Seq[Block], totalDifficulties: Seq[BigInt]): Unit = {
+    blocks.zip(totalDifficulties).foreach {
+      case (block, td) =>
+        broadcaster.broadcastBlock(NewBlock(block, td), handshakedPeers)
+    }
+  }
+
+  private def updateTxAndOmmerPools(blocksAdded: Seq[Block], blocksRemoved: Seq[Block]): Unit = {
+    blocksRemoved.headOption.foreach(block => ommersPool ! AddOmmers(block.header))
+    blocksRemoved.foreach(block => pendingTransactionsManager ! AddTransactions(block.body.transactionList.toSet))
+
+    blocksAdded.foreach { block =>
+      ommersPool ! RemoveOmmers(block.header :: block.body.uncleNodesList.toList)
+      pendingTransactionsManager ! RemoveTransactions(block.body.transactionList)
+    }
+  }
+
+  private def importWith(state: ImporterState, importFuture: => Future[Behavior]): Unit = {
+    val newState = ImporterState.importingBlocks(state)
+    context become running(ImporterState.importingBlocks(state))
+    importFuture.onComplete {
+      case Failure(ex) => throw ex
+      case Success(behavior) => context become behavior(ImporterState.notImportingBlocks(newState))
+    }
+  }
+}
+object BlockImporter {
+  def props(fetcher: ActorRef, ledger: Ledger, blockchain: Blockchain, syncConfig: SyncConfig): Props =
+    Props(new BlockImporter(fetcher, ledger, blockchain, syncConfig))
+
+  type Behavior = ImporterState => Receive
+
+  sealed trait ImporterMsg
+  case object Start extends ImporterMsg
+  case object OnTop extends ImporterMsg
+  case object NotOnTop extends ImporterMsg
+  case class MinedBlock(block: Block) extends ImporterMsg
+  case class ImportNewBlock(block: Block, peerId: PeerId) extends ImporterMsg
+
+  case class ImporterState(isOnTop: Boolean, importing: Boolean)
+  object ImporterState {
+    val initial: ImporterState = ImporterState(isOnTop = false, importing = false)
```

Comment:
The same thing. It could be much simpler if you move these methods to case class